File: bot/highlight_clipper/event_detector.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from .config import MULTIKILL_MIN_KILLS
from .config import MULTIKILL_THRESHOLD_SECONDS
from .config import TEAMFIGHT_MIN_KILLS
from .config import TEAMFIGHT_THRESHOLD_SECONDS

_CLOSE_FIGHT_WINDOW_S = 40
_MAX_PRE_ROLL_S = 35  # Clip-Vorlauf nie länger als das


@dataclass(slots=True)
class HighlightEvent:
    event_type: Literal["multikill", "teamfight", "close_fight"]
    game_time_s: int
    duration_s: int
    kill_count: int
    label: str
    pre_roll_s: int = field(default=0)  # dynamischer Vorlauf basierend auf TTK


def detect_events(account_id: int, match_info: dict) -> list[HighlightEvent]:
    players = [p for p in (match_info.get("players") or []) if isinstance(p, dict)]
    player_slot = _find_player_slot(account_id, players)
    if player_slot is None:
        return []

    all_deaths = _collect_deaths(players)
    all_deaths.sort(key=lambda d: d["game_time_s"])

    player_kills = [d for d in all_deaths if d["killer_player_slot"] == player_slot]
    player_own_deaths = [d for d in all_deaths if d["killed_player_slot"] == player_slot]

    events: list[HighlightEvent] = []

    # 1) Multi-kills
    player_kills_sorted = sorted(player_kills, key=lambda d: d["game_time_s"])
    multikill_ranges = _find_multikill_ranges(player_kills_sorted)
    for start, end in multikill_ranges:
        kills = player_kills_sorted[start:end]
        kill_count = len(kills)
        first_t = kills[0]["game_time_s"]
        last_t = kills[-1]["game_time_s"]
        max_ttk = max((k.get("time_to_kill_s") or 0) for k in kills)
        pre_roll = min(int(max_ttk) + 5, _MAX_PRE_ROLL_S)
        events.append(HighlightEvent(
            event_type="multikill",
            game_time_s=first_t,
            duration_s=last_t - first_t,
            kill_count=kill_count,
            label=f"{_multikill_name(kill_count)} ({kill_count} Kills)",
            pre_roll_s=pre_roll,
        ))

    # 2) Teamfights (verkettete Todesfälle)
    for fight in _find_teamfights(all_deaths, player_slot):
        first_t = fight[0]["game_time_s"]
        last_t = fight[-1]["game_time_s"]
        player_fight_deaths = [
            d for d in fight
            if d["killer_player_slot"] == player_slot or d["killed_player_slot"] == player_slot
        ]
        max_ttk = max((d.get("time_to_kill_s") or 0) for d in player_fight_deaths) if player_fight_deaths else 0
        pre_roll = min(int(max_ttk) + 5, _MAX_PRE_ROLL_S)
        events.append(HighlightEvent(
            event_type="teamfight",
            game_time_s=first_t,
            duration_s=last_t - first_t,
            kill_count=len(fight),
            label=f"Team Fight ({len(fight)} Kills)",
            pre_roll_s=pre_roll,
        ))

    # 3) Close fights: Kill + eigener Tod innerhalb CLOSE_FIGHT_WINDOW_S
    for close in _find_close_fights(player_kills_sorted, player_own_deaths):
        events.append(close)

    events = _deduplicate_events(events)
    return sorted(events, key=lambda e: e.game_time_s)


def _find_close_fights(
    player_kills: list[dict],
    player_own_deaths: list[dict],
) -> list[HighlightEvent]:
    results: list[HighlightEvent] = []
    used_kill_ids: set[int] = set()
    used_death_ids: set[int] = set()

    for kill in player_kills:
        kill_t = kill["game_time_s"]
        kill_ttk = kill.get("time_to_kill_s") or 0
        for death in player_own_deaths:
            death_t = death["game_time_s"]
            death_ttk = death.get("time_to_kill_s") or 0
            if abs(kill_t - death_t) > _CLOSE_FIGHT_WINDOW_S:
                continue
            if id(kill) in used_kill_ids or id(death) in used_death_ids:
                continue

            first_t = min(kill_t, death_t)
            last_t = max(kill_t, death_t)
            max_ttk = max(kill_ttk, death_ttk)
            pre_roll = min(int(max_ttk) + 5, _MAX_PRE_ROLL_S)
            label = "Clutch Kill" if kill_t > death_t else "Close Fight"

            results.append(HighlightEvent(
                event_type="close_fight",
                game_time_s=first_t,
                duration_s=last_t - first_t,
                kill_count=1,
                label=label,
                pre_roll_s=pre_roll,
            ))
            used_kill_ids.add(id(kill))
            used_death_ids.add(id(death))
            break

    return results


def _deduplicate_events(events: list[HighlightEvent]) -> list[HighlightEvent]:
    result: list[HighlightEvent] = []
    for e in sorted(events, key=lambda x: (x.game_time_s, -x.duration_s)):
        e_start = e.game_time_s - e.pre_roll_s
        e_end = e.game_time_s + e.duration_s
        dominated = False
        for kept in result:
            k_start = kept.game_time_s - kept.pre_roll_s
            k_end = kept.game_time_s + kept.duration_s
            if k_start <= e_start and k_end >= e_end:
                dominated = True
                break
        if not dominated:
            result.append(e)
    return result


def _find_player_slot(account_id: int, players: list[dict]) -> int | None:
    for player in players:
        if _as_int(player.get("account_id")) == account_id:
            return _as_int(player.get("player_slot"))
    return None


def _collect_deaths(players: list[dict]) -> list[dict]:
    deaths: list[dict] = []
    for player in players:
        slot = _as_int(player.get("player_slot"))
        for death in player.get("death_details") or []:
            if not isinstance(death, dict):
                continue
            game_time_s = _as_int(death.get("game_time_s"))
            if game_time_s is None:
                continue
            deaths.append({
                "game_time_s": game_time_s,
                "killer_player_slot": _as_int(death.get("killer_player_slot")),
                "killed_player_slot": slot,
                "time_to_kill_s": death.get("time_to_kill_s"),
            })
    return deaths


def _find_multikill_ranges(player_kills: list[dict]) -> list[tuple[int, int]]:
    ranges: list[tuple[int, int]] = []
    start = 0
    while start < len(player_kills):
        end = start + 1
        while end < len(player_kills):
            if player_kills[end]["game_time_s"] - player_kills[start]["game_time_s"] > MULTIKILL_THRESHOLD_SECONDS:
                break
            end += 1
        if end - start >= MULTIKILL_MIN_KILLS:
            ranges.append((start, end))
            start = end
            continue
        start += 1
    return ranges


def _find_teamfights(all_deaths: list[dict], player_slot: int) -> list[list[dict]]:
    """Verkettete Teamfight-Erkennung: Tode die ≤ THRESHOLD_S auseinanderliegen gehören zusammen."""
    if not all_deaths:
        return []

    fights: list[list[dict]] = []
    current: list[dict] = [all_deaths[0]]

    for death in all_deaths[1:]:
        if death["game_time_s"] - current[-1]["game_time_s"] <= TEAMFIGHT_THRESHOLD_SECONDS:
            current.append(death)
        else:
            _maybe_add_fight(current, player_slot, fights)
            current = [death]

    _maybe_add_fight(current, player_slot, fights)
    return fights


def _maybe_add_fight(deaths: list[dict], player_slot: int, out: list) -> None:
    player_kills = sum(1 for d in deaths if d["killer_player_slot"] == player_slot)
    if len(deaths) >= TEAMFIGHT_MIN_KILLS and player_kills >= 1:
        out.append(deaths)


def _multikill_name(kill_count: int) -> str:
    return {2: "Double Kill", 3: "Triple Kill", 4: "Quadra Kill", 5: "Penta Kill"}.get(kill_count, "Multi Kill")


def _as_int(value: object) -> int | None:
    try:
        return int(value)
    except (TypeError, ValueError):
        return None

File: rust/crates/tb-highlight/src/event_detector.rs
//! API-basierte Highlight-Erkennung aus den Match-Metadaten der deadlock-api.
//!
//! Port von `bot/highlight_clipper/event_detector.py`. Definiert den geteilten
//! [`HighlightEvent`]-Typ (auch vom demo-basierten Pfad genutzt) und erkennt
//! Multikills, Teamfights und Close-Fights aus `match_info["players"]`. Rein
//! und ohne I/O — vollständig deterministisch testbar.

use crate::config::{
    MULTIKILL_MIN_KILLS, MULTIKILL_THRESHOLD_SECONDS, TEAMFIGHT_MIN_KILLS,
    TEAMFIGHT_THRESHOLD_SECONDS,
};

const CLOSE_FIGHT_WINDOW_S: i64 = 40;
const MAX_PRE_ROLL_S: i64 = 35;

/// Art eines Highlight-Events.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EventType {
    Multikill,
    Teamfight,
    CloseFight,
}

impl EventType {
    /// String-Repräsentation wie im Python-`Literal` (für Worker/Serialisierung).
    pub fn as_str(self) -> &'static str {
        match self {
            EventType::Multikill => "multikill",
            EventType::Teamfight => "teamfight",
            EventType::CloseFight => "close_fight",
        }
    }
}

/// Ein erkanntes Highlight-Fenster für den Clip-Worker.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HighlightEvent {
    pub event_type: EventType,
    pub game_time_s: i64,
    pub duration_s: i64,
    pub kill_count: i64,
    pub label: String,
    /// Dynamischer Vorlauf basierend auf Time-to-Kill.
    pub pre_roll_s: i64,
}

/// Ein Todesfall aus den Metadaten (intern).
#[derive(Debug, Clone)]
struct Death {
    game_time_s: i64,
    killer_player_slot: Option<i64>,
    killed_player_slot: Option<i64>,
    time_to_kill_s: Option<f64>,
}

/// Erkennt Highlight-Events des Spielers (`account_id`) aus den Match-Metadaten.
pub fn detect_events(account_id: i64, match_info: &serde_json::Value) -> Vec<HighlightEvent> {
    let players: Vec<&serde_json::Value> = match_info
        .get("players")
        .and_then(serde_json::Value::as_array)
        .map(|a| a.iter().filter(|p| p.is_object()).collect())
        .unwrap_or_default();

    let player_slot = match find_player_slot(account_id, &players) {
        Some(slot) => slot,
        None => return Vec::new(),
    };

    let mut all_deaths = collect_deaths(&players);
    all_deaths.sort_by_key(|d| d.game_time_s);

    let player_kills: Vec<&Death> = all_deaths
        .iter()
        .filter(|d| d.killer_player_slot == Some(player_slot))
        .collect();
    let player_own_deaths: Vec<&Death> = all_deaths
        .iter()
        .filter(|d| d.killed_player_slot == Some(player_slot))
        .collect();

    let mut events: Vec<HighlightEvent> = Vec::new();

    // 1) Multikills
    for (start, end) in find_multikill_ranges(&player_kills) {
        let kills = &player_kills[start..end];
        let kill_count = kills.len() as i64;
        let first_t = kills[0].game_time_s;
        let last_t = kills[kills.len() - 1].game_time_s;
        let max_ttk = max_ttk(kills.iter().copied());
        let pre_roll = ((max_ttk as i64) + 5).min(MAX_PRE_ROLL_S);
        events.push(HighlightEvent {
            event_type: EventType::Multikill,
            game_time_s: first_t,
            duration_s: last_t - first_t,
            kill_count,
            label: format!("{} ({} Kills)", multikill_name(kill_count), kill_count),
            pre_roll_s: pre_roll,
        });
    }

    // 2) Teamfights (verkettete Todesfälle)
    for fight in find_teamfights(&all_deaths, player_slot) {
        let first_t = fight[0].game_time_s;
        let last_t = fight[fight.len() - 1].game_time_s;
        let player_fight_deaths = fight.iter().copied().filter(|d| {
            d.killer_player_slot == Some(player_slot) || d.killed_player_slot == Some(player_slot)
        });
        let max_ttk = max_ttk(player_fight_deaths);
        let pre_roll = ((max_ttk as i64) + 5).min(MAX_PRE_ROLL_S);
        events.push(HighlightEvent {
            event_type: EventType::Teamfight,
            game_time_s: first_t,
            duration_s: last_t - first_t,
            kill_count: fight.len() as i64,
            label: format!("Team Fight ({} Kills)", fight.len()),
            pre_roll_s: pre_roll,
        });
    }

    // 3) Close-Fights: Kill + eigener Tod innerhalb CLOSE_FIGHT_WINDOW_S
    events.extend(find_close_fights(&player_kills, &player_own_deaths));

    let events = deduplicate_events(events);
    let mut events = events;
    events.sort_by_key(|e| e.game_time_s);
    events
}

/// Maximale Time-to-Kill über eine Death-Sequenz (`ttk or 0`); leer → 0.
fn max_ttk<'a>(deaths: impl Iterator<Item = &'a Death>) -> f64 {
    deaths
        .map(|d| d.time_to_kill_s.unwrap_or(0.0))
        .fold(None::<f64>, |acc, v| Some(acc.map_or(v, |a| a.max(v))))
        .unwrap_or(0.0)
}

fn find_close_fights(
    player_kills: &[&Death],
    player_own_deaths: &[&Death],
) -> Vec<HighlightEvent> {
    let mut results = Vec::new();
    let mut used_death_idx = std::collections::HashSet::new();

    // Jeder Kill wird im äußeren Loop genau einmal betrachtet (break nach Match),
    // daher genügt ein used-Set für die Tode (used_kill_ids in Python redundant).
    for kill in player_kills {
        let kill_t = kill.game_time_s;
        let kill_ttk = kill.time_to_kill_s.unwrap_or(0.0);
        for (di, death) in player_own_deaths.iter().enumerate() {
            let death_t = death.game_time_s;
            let death_ttk = death.time_to_kill_s.unwrap_or(0.0);
            if (kill_t - death_t).abs() > CLOSE_FIGHT_WINDOW_S {
                continue;
            }
            if used_death_idx.contains(&di) {
                continue;
            }
            let first_t = kill_t.min(death_t);
            let last_t = kill_t.max(death_t);
            let max_ttk = kill_ttk.max(death_ttk);
            let pre_roll = ((max_ttk as i64) + 5).min(MAX_PRE_ROLL_S);
            let label = if kill_t > death_t { "Clutch Kill" } else { "Close Fight" };
            results.push(HighlightEvent {
                event_type: EventType::CloseFight,
                game_time_s: first_t,
                duration_s: last_t - first_t,
                kill_count: 1,
                label: label.to_string(),
                pre_roll_s: pre_roll,
            });
            used_death_idx.insert(di);
            break;
        }
    }
    results
}

fn deduplicate_events(events: Vec<HighlightEvent>) -> Vec<HighlightEvent> {
    let mut sorted = events;
    // Sortierung wie Python: (game_time_s aufsteigend, duration_s absteigend).
    sorted.sort_by(|a, b| {
        a.game_time_s
            .cmp(&b.game_time_s)
            .then(b.duration_s.cmp(&a.duration_s))
    });
    let mut result: Vec<HighlightEvent> = Vec::new();
    for e in sorted {
        let e_start = e.game_time_s - e.pre_roll_s;
        let e_end = e.game_time_s + e.duration_s;
        let dominated = result.iter().any(|kept| {
            let k_start = kept.game_time_s - kept.pre_roll_s;
            let k_end = kept.game_time_s + kept.duration_s;
            k_start <= e_start && k_end >= e_end
        });
        if !dominated {
            result.push(e);
        }
    }
    result
}

fn find_player_slot(account_id: i64, players: &[&serde_json::Value]) -> Option<i64> {
    for player in players {
        if as_int(player.get("account_id")) == Some(account_id) {
            return as_int(player.get("player_slot"));
        }
    }
    None
}

fn collect_deaths(players: &[&serde_json::Value]) -> Vec<Death> {
    let mut deaths = Vec::new();
    for player in players {
        let slot = as_int(player.get("player_slot"));
        let Some(details) = player.get("death_details").and_then(serde_json::Value::as_array) else {
            continue;
        };
        for death in details {
            if !death.is_object() {
                continue;
            }
            let Some(game_time_s) = as_int(death.get("game_time_s")) else {
                continue;
            };
            deaths.push(Death {
                game_time_s,
                killer_player_slot: as_int(death.get("killer_player_slot")),
                killed_player_slot: slot,
                time_to_kill_s: death.get("time_to_kill_s").and_then(serde_json::Value::as_f64),
            });
        }
    }
    deaths
}

fn find_multikill_ranges(player_kills: &[&Death]) -> Vec<(usize, usize)> {
    let mut ranges = Vec::new();
    let mut start = 0;
    while start < player_kills.len() {
        let mut end = start + 1;
        while end < player_kills.len() {
            if player_kills[end].game_time_s - player_kills[start].game_time_s
                > MULTIKILL_THRESHOLD_SECONDS
            {
                break;
            }
            end += 1;
        }
        if end - start >= MULTIKILL_MIN_KILLS {
            ranges.push((start, end));
            start = end;
            continue;
        }
        start += 1;
    }
    ranges
}

fn find_teamfights<'a>(all_deaths: &'a [Death], player_slot: i64) -> Vec<Vec<&'a Death>> {
    if all_deaths.is_empty() {
        return Vec::new();
    }
    let mut fights: Vec<Vec<&Death>> = Vec::new();
    let mut current: Vec<&Death> = vec![&all_deaths[0]];

    for death in &all_deaths[1..] {
        if death.game_time_s - current[current.len() - 1].game_time_s <= TEAMFIGHT_THRESHOLD_SECONDS
        {
            current.push(death);
        } else {
            maybe_add_fight(&current, player_slot, &mut fights);
            current = vec![death];
        }
    }
    maybe_add_fight(&current, player_slot, &mut fights);
    fights
}

fn maybe_add_fight<'a>(deaths: &[&'a Death], player_slot: i64, out: &mut Vec<Vec<&'a Death>>) {
    let player_kills = deaths
        .iter()
        .filter(|d| d.killer_player_slot == Some(player_slot))
        .count();
    if deaths.len() >= TEAMFIGHT_MIN_KILLS && player_kills >= 1 {
        out.push(deaths.to_vec());
    }
}

/// Name eines Multikills nach Kill-Zahl (Python `_multikill_name`).
pub fn multikill_name(kill_count: i64) -> &'static str {
    match kill_count {
        2 => "Double Kill",
        3 => "Triple Kill",
        4 => "Quadra Kill",
        5 => "Penta Kill",
        _ => "Multi Kill",
    }
}

/// `int(value)`-Semantik aus Python: Zahlen werden (gegen 0) gekürzt, Strings
/// nur als reine Ganzzahl geparst (`int("12.5")` schlägt fehl → None).
fn as_int(value: Option<&serde_json::Value>) -> Option<i64> {
    match value {
        Some(serde_json::Value::Number(n)) => n.as_i64().or_else(|| n.as_f64().map(|f| f as i64)),
        Some(serde_json::Value::String(s)) => s.trim().parse::<i64>().ok(),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn multikill_namen() {
        assert_eq!(multikill_name(2), "Double Kill");
        assert_eq!(multikill_name(5), "Penta Kill");
        assert_eq!(multikill_name(6), "Multi Kill");
    }

    #[test]
    fn as_int_python_semantik() {
        assert_eq!(as_int(Some(&json!(7))), Some(7));
        assert_eq!(as_int(Some(&json!(7.9))), Some(7)); // Truncation
        assert_eq!(as_int(Some(&json!("123"))), Some(123));
        assert_eq!(as_int(Some(&json!("12.5"))), None); // int("12.5") wirft
        assert_eq!(as_int(Some(&json!(null))), None);
        assert_eq!(as_int(None), None);
    }

    #[test]
    fn detect_events_multikill_dreierserie() {
        let match_info = json!({
            "players": [
                {"account_id": 100, "player_slot": 0, "death_details": []},
                {"account_id": 200, "player_slot": 1, "death_details": [
                    {"game_time_s": 100, "killer_player_slot": 0, "time_to_kill_s": 3}]},
                {"account_id": 300, "player_slot": 2, "death_details": [
                    {"game_time_s": 105, "killer_player_slot": 0, "time_to_kill_s": 2}]},
                {"account_id": 400, "player_slot": 3, "death_details": [
                    {"game_time_s": 108, "killer_player_slot": 0}]},
            ]
        });
        let events = detect_events(100, &match_info);
        assert_eq!(events.len(), 1);
        let e = &events[0];
        assert_eq!(e.event_type, EventType::Multikill);
        assert_eq!(e.label, "Triple Kill (3 Kills)");
        assert_eq!(e.game_time_s, 100);
        assert_eq!(e.duration_s, 8);
        assert_eq!(e.pre_roll_s, 8); // int(max_ttk=3)+5
    }

    #[test]
    fn detect_events_unbekannter_spieler_leer() {
        let match_info = json!({"players": [{"account_id": 1, "player_slot": 0}]});
        assert!(detect_events(999, &match_info).is_empty());
    }

    #[test]
    fn detect_events_teamfight_ab_vier_toden() {
        // 4 verkettete Tode, Spieler (slot 0) hat ≥1 Kill → Teamfight.
        let match_info = json!({
            "players": [
                {"account_id": 100, "player_slot": 0, "death_details": [
                    {"game_time_s": 210, "killer_player_slot": 5}]},
                {"account_id": 200, "player_slot": 1, "death_details": [
                    {"game_time_s": 200, "killer_player_slot": 0}]},
                {"account_id": 300, "player_slot": 2, "death_details": [
                    {"game_time_s": 205, "killer_player_slot": 0}]},
                {"account_id": 400, "player_slot": 3, "death_details": [
                    {"game_time_s": 208, "killer_player_slot": 5}]},
            ]
        });
        let events = detect_events(100, &match_info);
        let teamfights: Vec<_> = events
            .iter()
            .filter(|e| e.event_type == EventType::Teamfight)
            .collect();
        assert_eq!(teamfights.len(), 1);
        assert_eq!(teamfights[0].kill_count, 4);
        assert_eq!(teamfights[0].label, "Team Fight (4 Kills)");
    }

    #[test]
    fn close_fight_clutch_vs_close() {
        // Kill nach eigenem Tod → "Clutch Kill"; Kill vor Tod → "Close Fight".
        let kill = Death { game_time_s: 50, killer_player_slot: Some(0), killed_player_slot: Some(1), time_to_kill_s: Some(4.0) };
        let own_death = Death { game_time_s: 40, killer_player_slot: Some(9), killed_player_slot: Some(0), time_to_kill_s: Some(2.0) };
        let kills = vec![&kill];
        let deaths = vec![&own_death];
        let out = find_close_fights(&kills, &deaths);
        assert_eq!(out.len(), 1);
        assert_eq!(out[0].label, "Clutch Kill"); // kill_t 50 > death_t 40
        assert_eq!(out[0].game_time_s, 40);
        assert_eq!(out[0].duration_s, 10);
        assert_eq!(out[0].pre_roll_s, 9); // int(max(4,2))+5
    }

    #[test]
    fn dedup_dominierte_events_entfernt() {
        let outer = HighlightEvent { event_type: EventType::Teamfight, game_time_s: 100, duration_s: 30, kill_count: 4, label: "A".into(), pre_roll_s: 10 };
        // inner liegt komplett im Fenster von outer → wird entfernt.
        let inner = HighlightEvent { event_type: EventType::Multikill, game_time_s: 105, duration_s: 5, kill_count: 2, label: "B".into(), pre_roll_s: 0 };
        let out = deduplicate_events(vec![outer.clone(), inner]);
        assert_eq!(out.len(), 1);
        assert_eq!(out[0].label, "A");
    }
}
